### DISCUSSION TITLE
Sanitize and sanitizer

### DISCUSSION THREAD

**Discussion by joshbruce** [markedjs/marked#1232] [General]
**Marked version:** 0.3.19

**Markdown flavor:** n/a

**Proposal type:** other

## What pain point are you perceiving?

Marked is taking on too much; or, at least more than it needs to.

## What solution are you suggesting?

It appears at some point someone had the idea of adding sanitization to marked. Later, someone else had the idea of letting people plug in their own sanitizer.

I suggest we go for the latter and deprecate the former. Turning on sanitize doesn't really seem to give us much in all honesty.

**Comment by UziTech** [comment]
I say we remove sanitizing all together and recommend `DOMPurify.sanitize(marked(...))`

**Comment by Martii** [comment]
I'm +1 for using any external sanitizers only as we've never really used the internal one here. [Even in the DOM](https://github.com/OpenUserJS/OpenUserJS.org/blob/fdcfc7a87077b0c78d773aab49f194885d787194/views/includes/scripts/markdownEditor.html#L14) we don't utilize it e.g. using `false` *(technically `undefined` on our end but [doc'd currently as `false`](https://marked.js.org/#/USING_ADVANCED.md) )* default to my knowledge. The one we use is highly configurable and the wheel there is well polished imho.

Be aware that this change could affect other packages so strong announcements and CHANGELOGs are highly appreciated *(and of course semver announcements too)*. :)

**Comment by joshbruce** [comment]
@Martii: Thanks for chiming in as well. Right now we don't have a direct way to contact users. I was hoping more people would have added themselves to the "users" portion of the Authors page, to be honest. So, we're counting on the tickets, milestones, and projects to give people the warning.

We'll probably spend most of 1.x preparing people to lose things.

@UziTech: Agreed.

Tagging #1114

**Comment by alystair** [comment]
If you're discussing escaping HTML... here's a solution that utilizes a browsers own filtering. If it breaks it means there's a browser vulnerability and you made yourself some nice pocket change via their  vulnerability rewards program.
```js
var escape = document.createElement('textarea');
function escapeHTML(html) { escape.textContent=html; return escape.innerHTML; }
function unescapeHTML(html) { escape.innerHTML=html; return escape.textContent; }
```
This doesn't solve the XSS vuln in the comment above but does for plenty of other things, assuming one doesn't want HTML code passthrough :)

**Comment by styfle** [comment]
Another 3rd party sanitizer I just found: https://github.com/rhysd/marked-sanitizer-github

I can't say how secure it is but the point remains that marked doesn't need to ship with a sanitizer.

-----

*Update*: let's make a list of popular sanitizers so our users don't need to search the millions of packages and then they can make a decision:

- `dompurify`: https://www.npmjs.com/package/dompurify
- `sanitize-html`: https://www.npmjs.com/package/sanitize-html
- `insane`: https://www.npmjs.com/package/insane
- `marked-sanitizer-github`: https://www.npmjs.com/package/marked-sanitizer-github

➡️ See my response in [#1388](https://github.com/markedjs/marked/issues/1388#issuecomment-447048317) for example code.

**Comment by OrKoN** [comment]
I think it's nice if marked has sanitization built-in.

I tried to use `dompurify` but I find it too heavy and slow for Lambda environment/nodejs. I have tried `sanitize-html` but it seems to be simply removing everything that is not allowed and strips class names produced by highlight.js so requires a lot of tuning/testing.

**Reply by tipiirai** [reply]
Agreed! dompurify is really not a good solution to the problem. It added 12M of extra clutter to node_modules and my application start time is now significantly slower.

**Reply by UziTech** [reply]
That is the reason we don't have sanitation built it. In order to do it right we would need to maintain a lot more code and things would go much slower. I'm pretty sure we wouldn't be able to do a better job on either of those fronts than dompurify or any other package focused on sanitizing HTML.

**Comment by UziTech** [comment]
We will most likely have a plug-in to do sanitization with a third party library in the future to make it easy for people to use, but it doesn't make sense for core marked to have sanitization built in because it is not part of the markdown specs.

**Comment by UziTech** [comment]
As a matter of fact the common mark spec explicitly allows unsanitized html
[Example #138](https://spec.commonmark.org/0.28/#example-138)

**Comment by OrKoN** [comment]
A plugin would be very nice. I am not an expert in XSS prevention and HTML sanitization (and I feel many other users aren't too) so it would be awesome if there is a standard/recommended way to do it. I found some articles online with examples like this one https://shuheikagawa.com/blog/2015/09/21/using-highlight-js-with-marked/ which seem to open up a possibility for XSS. Perhaps the plugin can be included by default (or in the getting started guide) because I am afraid many users don't realize the risk and apply the same marked configuration to untrusted data.

**Comment by OrKoN** [comment]
I wrote a benchmark to evaluate different sanitizers https://github.com/OrKoN/marked-benchmark I got some interesting results:

```
  builtInSanitizer   x 221 ops/sec ±1.48% (85 runs sampled)
  domPurifySanitizer x 148 ops/sec ±2.85% (78 runs sampled)
  noSanitizer        x 216 ops/sec ±0.62% (83 runs sampled)
```

So with dom purify it seems to be roughly twice slower for me compared to the built-in sanitizer.

P.S. updated the results. The previous message contained the wrong ones.
P.P.S. it'd be nice to create instances of marked
P.P.P.S. the code tag renderer seems to be always getting undefined as `escaped` param https://github.com/markedjs/marked/blob/master/lib/marked.js#L1200 no matter what the value of sanitize flag is
P.P.P.P.S. it looks like dom purify does not escape but simply removes not allowed tags

**Comment by OrKoN** [comment]
Another 3rd party sanitizer which looks suitable https://github.com/bevacqua/insane

**Comment by lionel-rowe** [comment]
FWIW, I just tried plugging in a third-party sanitizer (insane) for use with the `html` method of the renderer, and I would **strongly recommend _against_ this approach**. There are a lot of edge cases for which user-supplied HTML isn't recognized as such, especially if it's inline with some markdown or the last line of the input. Example:

```js
const renderer = new marked.Renderer();
renderer.html = html => insane(html);

marked.setOptions({ renderer });

marked('<img src="sdfg" onerror="alert(1)">');
// expected: '<p><img src="sdfg"></p>'
// actual:   '<p><img src="sdfg" onerror="alert(1)"></p>'
```

It seems like the only safe way of doing this is passing the whole rendered output through the chosen third-party sanitizer, i.e. first set whitelist options that play nicely with your markdown setup ([example for insane + marked + hljs](https://gist.github.com/lionel-rowe/bb384465ba4e4c81a9c8dada84167225)), then use as a wrapper like this:

```js
insane(marked(input), insaneOptions);
```

**Comment by OrKoN** [comment]
@lionel-rowe I am wondering what is your use case? Because for me the expected result for the img case would be: `<p>&lt;img src=&quot;sdfg&quot; onerror=&quot;alert(1)&quot; /&gt;</p>` and not `<p><img src="sdfg"></p>` I am looking for the solution to keep what is not allowed in the output but escape it instead of removing it? So far I see this behavior only with the marked's `sanitize=true` option.

**Comment by OrKoN** [comment]
@lionel-rowe I have added your config for `insane` to my [benchmark repo](https://github.com/OrKoN/marked-benchmark) 

Does anyone have a good config for `DOMPurify` for the makred + hljs use case?

**Comment by styfle** [comment]
@OrKoN See [this DOMPurify comment](https://github.com/markedjs/marked/issues/1388#issuecomment-447048317) for example usage.

**Comment by lionel-rowe** [comment]
> @lionel-rowe I am wondering what is your use case? Because for me the expected result for the img case would be: `<p>&lt;img src=&quot;sdfg&quot; onerror=&quot;alert(1)&quot; /&gt;</p>` and not `<p><img src="sdfg"></p>` I am looking for the solution to keep what is not allowed in the output but escape it instead of removing it? So far I see this behavior only with the marked's `sanitize=true` option.

My use case is that I want to allow user-supplied HTML to be parsed as HTML, but in a safe way. It's not especially useful in the case of an `img` tag or other things already supported by Markdown, but it's nice to have for things such as `kbd` and `details`/`summary`.

As insane is whitelist-based and simply drops nonconforming elements/attributes, I'm guessing it wouldn't be ideal for your use case. Not sure if any of the other third-party sanitizers would be suitable for this.

**Comment by OrKoN** [comment]
@styfle thanks, but it looks like an example for `sanitize-html` and not for the `DOMPurify`.

So it seems to me that there are two approaches to sanitization:

1. stripping all HTML constructs which are not allowed (as per a whitelist). This approach seems to be implemented by all sanitizers mentioned in this thread. Also, this approach does not make a distinction (obviously) between HTML emitted by marked and HTML that is a part of the markdown content (see the comment by @lionel-rowe above). 

2. escaping HTML constructs so that they cannot be interpreted as HTML. It seems that this approach is what the marked'd flag `sanitize=true` currently does. Since marked knows what part of HTML it emits, it can easily escape all other output so that the user content is not treated as HTML. It seems to be impossible to achieve this functionality using a 3rd party sanitizers mentioned so far because they don't escape HTML.

The second options looks pretty much like just markdown but any custom HTML is disallowed. I guess the safest option would be to escape all content and then pass the entire marked output to a stand-alone sanitizer. So something like this would work:

```
const renderer = new EscapeRenderer(); // renderer same as the current one with sanitize=true 
marked.setOptions({ renderer });
insane(marked(input), insaneOptions);
```

Perhaps there are existing libs which do HTML escaping. Right now the escaping is implemented in the marked's core.

**Comment by styfle** [comment]
> but it looks like an example for sanitize-html and not for the DOMPurify.

Oops! You are correct 😅 

> Also, this [whitelist] approach does not make a distinction between HTML emitted by marked and HTML that is a part of the markdown content (see the comment by @lionel-rowe above).

I don't see how the requirements from @lionel-rowe need to make this distinction. To me, this can still be implemented as a whitelist. Either you are allowing `img` or you aren't and you can use something like `html-sanitize` to transform attributes whether the user wrote `<img>` or `![img]()`.

To reiterate, I think you can solve your problem by running [sanitize-html](https://github.com/markedjs/marked/issues/1388#issuecomment-447048317) after `marked`.

Maybe you can elaborate more on why a whitelist approach doesn't work? Thanks!

**Comment by OrKoN** [comment]
I can give another example. The user enters Markdown content and does not know how it is sanitized, they want to see their content back. So the user enters

`Hi everyone! This is **how** you do XSS <script>alert('XSS')</script>`

Obviously, the `script` tag is not allowed by the whitelisting sanitizer, so the user gets back something like `<p>Hi everyone! This is <strong>how</stong> you do XSS </p>`. The part of the content was erased.

What would be desired instead is to escape that bad HTML, so that the user gets back:

`<p>Hi everyone! This is <strong>how</stong> you do XSS &lt;script&gt;alert(&apos;XSS&apos;)&lt;/script&gt;</p>`

Here it's challenging to know outside of markdown parsing context that `<script>` needs to be escaped but, for example, `<strong>` does not have to be. There was an issue opened about it in `sanitize-html` but it got closed without resolution https://github.com/punkave/sanitize-html/issues/57

Does it make sense?

So far I could not configure `sanitize-html` to escape instead of removing. It looks like this is not possible and is especially challenging because 3rd party sanitizer does not know what part of HTML was produced by marked and what part was the content.

**Comment by styfle** [comment]
@OrKoN You can solve this use case by doing the opposite&mdash;escape the html first and then run it through marked. This will prevent the user from writing any html and force them to use markdown to format text.

```js 
      const entityMap = {
        "&": "&amp;",
        "<": "&lt;",
        ">": "&gt;",
        '"': '&quot;',
        "'": '&#39;',
        "/": '&#x2F;'
      };

      const sanitize = (html) => {
        return html.replace(/[&<>"'\/]/g, key => entityMap[key]);
      }

      const dirty = "This is **how** you do XSS <script>alert('XSS')</" + "script>";
      const clean = sanitize(dirty);
      const html = marked(clean);
```

https://jsbin.com/vinocifoqi/edit?html,output

**Comment by OrKoN** [comment]
@styfle Thanks! It seems to work almost like expected, but I believe there will be edge cases when HTML entities are used for Markdown syntax. One example I can think of now is links: for example, `<https://google.com>`. It should be rendered as a link (`<p><a href="https://google.com">https://google.com</a></p>`) but with the escape everything approach it is displayed as is (`&lt;https://google.com&gt;`).

**Comment by styfle** [comment]
@OrKoN You got me 🤔 

Perhaps @UziTech can think of a solution for this use case 😄

**Comment by UziTech** [comment]
You could override the `renderer.html` to escape the html.

```js
const entityMap = {
  "&": "&amp;",
  "<": "&lt;",
  ">": "&gt;",
  '"': '&quot;',
  "'": '&#39;',
  "/": '&#x2F;'
};

const renderer = new marked.Renderer();
renderer.html = (html) => {
  return html.replace(/[&<>"'\/]/g, key => entityMap[key]);
}

const dirty = "This is **how** you do XSS <script>alert('XSS')</" + "script>";
const html = marked(dirty, {renderer});
```

The only other way I can think of is to escape html first as @styfle mentioned and try to catch as many of those edge cases as possible.

**Comment by OrKoN** [comment]
@UziTech Is the output safe in this case? I get `<p>This is <strong>how</strong> you do XSS <script>alert('XSS')</script></p>` which looks unsafe. The `<script>` tag is within the paragraph so the html method of the renderer is not called.

It seems to me that `sanitize=true` flag provides quite a valuable feature at the moment which is not easy to achieve by other means. By reading the code, it looks like turning `sanitize` to `true` changes how marked treats some tokens (e.g., HTML token seems to be treated as a paragraph), escapes encountered tags and also does something to sanitize the link content.

**Comment by UziTech** [comment]
@lionel-rowe it looks like you found a bug. The `renderer.html` doesn't get called because there isn't a new line (`\n`) at the end of the string so it doesn't think it is html

**Comment by UziTech** [comment]
@OrKoN according to the [commonmark spec](https://spec.commonmark.org/0.28/#html-block), html blocks must start on their own line otherwise it is just considered text. And the only thing that could catch it is sanitizing the text before or after it is parsed.

**Comment by OrKoN** [comment]
@UziTech I just started reading the spec recently, so I don't know it well. It looks like the spec does not define how to treat this case right? I mean it looks like it is correct that it is unescaped, but in practice, I'd say this is not always desirable.

**Comment by UziTech** [comment]
Sanitization is out of scope for markdown parsing, which is why I believe it should be done by a third party library.

**Comment by OrKoN** [comment]
The only question for me is how do it with a 3rd party library in the same way as marked does it currently? :thinking:

**Comment by UziTech** [comment]
this is essentially what marked does:

```js
const lexed = marked.lexer(dirty);
marked.parser(lexed.map(token => {return {...token, text: escape(token.text)};}));
```

**Comment by UziTech** [comment]
here is a working example

```js
const entityMap = {
  '&': '&amp;',
  '<': '&lt;',
  '>': '&gt;',
  '"': '&quot;',
  "'": '&#39;',
  '/': '&#x2F;'
};
const dirty = 'Hi everyone! This is **how** you do XSS <script>alert("XSS")</script>';
const lexed = marked.lexer(dirty);
const cleanLex = lexed.map(token => {
  token.text = token.text.replace(/[&<>"'\/]/g, key => entityMap[key]);
  return token;
});
cleanLex.links = lexed.links;
const cleanHtml = marked.parser(cleanLex);
// <p>Hi everyone! This is <strong>how</strong> you do XSS &lt;script&gt;alert(&quot;XSS&quot;)&lt;&#x2F;script&gt;</p>
```

**Comment by OrKoN** [comment]
@UziTech Thanks for the example! I think it's still missing something to sanitize links. Consider the following example `[some text](javascript:exec())`. It results in `<p><a href="javascript:exec()">some text</a></p>` whereas marked with the sanitize option returns `<p>some text</p>`. I guess one can re-define the renderer's link method to add checks in there but I guess the structure will be still `p -> a`.

**Comment by UziTech** [comment]
ya you would just need to override the `renderer.link` and `renderer.image` functions to sanitize the href.

**Comment by makeitcount** [comment]
@joshbruce 
**Do not support the move unless there is `htmlAllowed: false` option**
Given the less no. of people knowing markdown, I am taking a bet on Markdown only for the security purpose.

There are two cases I can see

1. People who do not want html rendering in content at all for security purpose(as in my case): They should not require any other library to do so
2. People who want to include html rendering in the content but want to make it as secure against XSS as possible: This should be outsourced to other libraries such as DOMPurify

The way I see it can be done is by having an option `htmlAllowed`, set to false by default.

P.S. I haven't gone in details of sanitize option yet if it is the same thing. Let me know if I'm missing something

**Comment by lunaru** [comment]
@UziTech I just wanted to chime in here to say that I agree with @OrKoN that there are 2 different concepts of "sanitize" that are being conflated. I fear that the move to remove the `sanitize` from the core of this library is missing out on one of the concepts.

`sanitize=true` today does not actually sanitize -- it actually escapes. E.g.:

```
marked("<div>foo</div>", { sanitize: true })
```

outputs 

```
<p>&lt;div&gt;foo&lt;/div&gt;</p>
```

This is desired when the user entered content should be escaped HTML. One nice thing about keeping this functionality in the core of marked is that marked is the only functional layer that understands what's output and what's input. Consider this:

```
marked("\\<div>foo</div>", { sanitize: true })
```

It's actually part of the Markdown spec to escape "<" with "\\" (backslash) to identify html elements. This should also output the same thing as the example from above.

None of the proposed solutions (DOMPurify, et al) solve this problem at all. I actually think this parameter should be renamed `escapeInput` and its functionality should be kept since it's actually core to recognizing and parsing markdown syntax.

TL;DR: `sanitize=true` operates on the input and is markdown dependent. Proposed solutions operate on the output and are html operations. These are not the same. Please reconsider removing `sanitize` functionality from core or just rename `sanitize` to reflect what it actually does.

**Comment by lunaru** [comment]
@koczkatamas I'd like your thoughts on the above comment as well, since it looks like the current deprecation warnings were in your commit

**Comment by koczkatamas** [comment]
@lunaru well, if you can solve escaping securely, then it's okay on my part, but I am not sure that the current (hardened) sanitize cannot be bypassed and used for XSS that's why I recommended deprecation.

**Comment by lunaru** [comment]
@koczkatamas I think the sanitize vs escape concept might be conflated. Why does escape need to be 100% secure when sanitize will ultimately "guarantee" proper sanitization? A proper secure pipeline looks like this: `rawText -> escape -> marked -> sanitize`. If a user cares about sanitization, they can use a sanitizer. But that's not what `sanitize=true` actually does. It does escaping, which doesn't need to be fully secure, but it does need to be core to marked, since escaping properly requires proper parsing of markdown syntax.

As an example, this is perfectly valid:

```
DOMPurify.sanitize(marked(..., { sanitize: true))
```

And it's *not* the same as this:

```
DOMPurify.sanitize(marked(..., { sanitize: false))
```

That's because the `sanitize` parameter should actually be called `escape`

**Comment by UziTech** [comment]
@lunaru escaping input and sanitizing output are separate tasks from converting markdown to html, which is why I believe it should be done separately.

> marked is the only functional layer that understands what's output and what's input

I agree that marked should allow extensions to access it's tokenized data and alter it like https://github.com/markedjs/marked/issues/1232#issuecomment-470704335

I would like marked to be more extensible so we can focus on the task of converting markdown to html according to the specs, and other developers can focus on other tasks (sanitizing, escaping, etc).

**Comment by lunaru** [comment]
@UziTech I don't disagree with the intention. However, I do think before removing the `sanitize` param,  a functional equivalent to `escape` (aka `sanitize=true`) should be discussed first. If that means altering tokenized data, that seems reasonable if documented. I'm just cautioning against putting the cart before the horse, where `sanitize` is removed without a suitable replacement. If you look at all of the existing suggestions (e.g. DOMPurify) they're not even addressing the right functionality, which concerns me, because `sanitize=true` is not a sanitizer, it's an escaper.

Also, I don't think the problem is that trivial. Consider the following, taken as an adaptation of your comment above:

```
var entityMap = {
  '&': '&amp;',
  '<': '&lt;',
  '>': '&gt;',
  '"': '&quot;',
  "'": '&#39;',
  '/': '&#x2F;'
};
var dirty = "\\<div>";
var lexed = marked.lexer(dirty);
var cleanLex = lexed.map(token => {
  token.text = token.text.replace(/[&<>"'\/]/g, key => entityMap[key]);
  return token;
});
cleanLex.links = lexed.links;
var cleanHtml = marked.parser(cleanLex);
```

Note that the output to `cleanHtml` is incorrect (it outputs `<p>&amp;lt;div&gt;</p>`, which incorrectly double encoded the ampersand) because it doesn't handle the escaped `\<div>` properly, but this works fine:

```
marked("\\<div>", { sanitize: true })
```

This outputs `<p>&lt;div&gt;</p>` which is the correct output since `\<div>` and `<div>` are equivalent when escaped.

**Comment by UziTech** [comment]
> before removing the sanitize param, a functional equivalent to escape (aka sanitize=true) should be discussed first.

I agree. The game plan right now is to get to 100% commonmark and gfm compliance then release v1.0 with (almost) all options deprecated, then make marked more extensible and remove options in v2.0

We will try to maintain a few extensions that function like each option we remove.

**Comment by A1rPun** [comment]
I agree sanitizing HTML is not the responsibility for marked. However since the deprecation notice I'm using this combination everywhere where I previously used only marked (browser apps):

```js
import marked from 'marked';
import DOMPurify from 'dompurify';

DOMPurify.sanitize(marked(myVar))
```

**Comment by azu** [comment]
Pluggable feature is good.
But, it is a bit difficult that use [DOMPurify](https://github.com/cure53/DOMPurify) as universal library. (It require to split entry point into browser and node.js)
The reason is that I've created [safe-marked](https://github.com/azu/safe-marked).

**Comment by UziTech** [comment]
That is exactly what we want to see, the community making extensions :100:. Ideally marked will handle converting markdown to html according to the spec and extensions will allow users to change the functionality.

We will hopefully be releasing v1.0 soon with a [new extension system](https://github.com/markedjs/marked/pull/1646) so extending marked will be easier in the future.

**Comment by azu** [comment]
@UziTech Nice!

**Comment by azu** [comment]
I've just created [marked-plugin-sanitizer](https://github.com/azu/marked-plugin-sanitizer) as PoC.
This new extension system looks like custom render system.
I found some problems in this [marked.use](https://marked.js.org/#/USING_PRO.md#use)

- Difficult to write post process like sanitizer
  - I've avoided this problem by wrapping default `Renderer`
  - https://github.com/azu/marked-plugin-sanitizer/blob/07018001396051795723387e0bcdef78841c57ef/src/SanitizeRender.ts#L10-L43
- A plugin can not get `marked`'s options, so user should pass `marked` option to the plugin.
  - But, I've noticed taht  `Rednder` respect `marked.setOptions` by default. 
  - https://github.com/azu/marked-plugin-sanitizer/blob/fcc7b586a0a6ad77ee0bd92a6b26e481ff1eb292/test/index.test.ts#L55-L64

So [marked-plugin-sanitizer](https://github.com/azu/marked-plugin-sanitizer) works, but it is tricky code.

**Comment by UziTech** [comment]
`marked.use` is only for modifying the marked options so multiple extensions can override the `renderer` and `tokenizer` without overriding each other.

If you are trying to do any pre-processing of the markdown or post-processing of the html your better off just having them send your package the markdown and run it through marked inside your package.

**Comment by Zemnmez** [comment]
a sanitizer is not really a panacaea for these kind of problems. it would be good to see marked use safe DOM operations (`.textContent`) to generate DOM nodes instead of raw, potentially unsafe HTML code

**Comment by UziTech** [comment]
@Zemnmez marked converts markdown to html not to DOM nodes. Although you could create a renderer that does that.
